File: 153. Find Minimum in Rotated Sorted Array/153. Find Minimum in Rotated Sorted Array.java
class Solution {
	public int findMin(int[] nums) {
   
	int start = 0;
    int end = nums.length -1;
    
        while (start <= end) {
            int mid = start + (end - start)/2;
			
            if (mid < end && nums[mid] > nums[mid + 1])
                return nums[mid + 1];     // greater index holds the min element in the array
            if (start < mid && nums[mid - 1] > nums[mid] )
                return nums[mid];     //  greater index holds the min element in the array
				
			if (nums[start] > nums[mid])    // search in the left half 
                end = mid - 1;
            else    // search in the right half
                start = mid + 1;
        } 
    return nums[0];
	}
}

File: 153. Find Minimum in Rotated Sorted Array/153. Find Minimum in Rotated Sorted Array.cpp
class Solution
{
public:
    int findMin(vector<int> &nums)
    {
        int start = 0;
        int end = nums.size() - 1;
        while (start < end)
        {
            int mid = start + (end - start) / 2;
            // right side is rotated, expecting smallest number between mid + 1 <=  end;
            if (nums[mid] > nums[end])
            {
                start = mid + 1;
            }
            else
            {
                // Right side is sequential, which  implies
                // left is either sequential or rotated. Either case, we can find minimum from values at between start <= mid index.
                // e.g.
                // 1 2 3 4  5 6 7   ( 4 is mid)
                // 8  9 4 5 6 7     (say 4 is mid)
                // 8 9 1 4 5 6 7
                end = mid;
            }
        }
        return nums[start];
    }
};